File: _08_02_/src/main/java/banco/dominio/Conta.java
package banco.dominio;

import banco.dominio.excecoes.ExcecaoChequeEspecial;

public class Conta {

    protected double saldo;

    public Conta(double saldoInicial) {
        this.saldo = saldoInicial;
    }

    public double getSaldo() {
        return saldo;
    }

    public boolean depositar(double valor) {
        saldo = saldo + valor;
        return true;
    }

    public void sacar(double valor) throws ExcecaoChequeEspecial {
        try {
            double deficit = this.saldo - valor;
            if (deficit < 0) {
                throw new ExcecaoChequeEspecial(deficit, "Saldo insuficiente");
            }
            this.saldo = deficit;
        } catch (ExcecaoChequeEspecial e) {
            throw e;
        }
    }


}

File: Trabalho_JS/banco/dominio/Conta.js
import ExcecaoChequeEspecial from "./excecoes/ExcecaoChequeEspecial.js";

export default class Conta {
    #numeroDaConta;
    _saldo;

    constructor(saldoInicial) {
        if (this.constructor instanceof Conta) {
            throw new Error("Object of Abstract class cannot be instantiated")
        }
        this._saldo = saldoInicial;
    }

    getNumeroDaConta() { return this.#numeroDaConta; }

    setNumeroDaConta(numeroDaConta) { this.#numeroDaConta = numeroDaConta; }

    getSaldo() { return this._saldo; }

    depositar(valor) {
        this._saldo += valor;
        return true; 
    }

    sacar(valor) {
        if ((this._saldo - valor) >= 0) {
            this._saldo -= valor;
         } else {
            let deficit = valor - this._saldo;
            throw new ExcecaoChequeEspecial(deficit, "Saldo insuficiente");
        }
    }
}